File: typespec/hub/auth/signup.ts
import type { NewPassword } from "../../common/authentication.ts";
import { isNewPassword, isOpaqueToken } from "../../common/authentication.ts";
import type { EmailAddress } from "../../common/common.ts";
import { isEmailAddress, normalizeEmailAddress } from "../../common/common.ts";
import type {
  CountryCode,
  DisplayName,
  FrontendLocale,
} from "../../common/localization.ts";
import {
  isCountryCode,
  isDisplayName,
  isFrontendLocale,
  normalizeDisplayName,
} from "../../common/localization.ts";
import type { HubHandle, HubUserDID } from "../types.ts";

export type HubSignupToken = string;

export interface RequestSignupRequest {
  email_address: EmailAddress;
  display_name: DisplayName;
  preferred_language: FrontendLocale;
  resident_country: CountryCode;
}

export function normalizeRequestSignupRequest(
  request: RequestSignupRequest,
): RequestSignupRequest {
  return {
    ...request,
    email_address: normalizeEmailAddress(request.email_address),
    display_name: normalizeDisplayName(request.display_name),
  };
}

export function validateRequestSignupRequest(
  request: RequestSignupRequest,
): string[] {
  const normalized = normalizeRequestSignupRequest(request);
  const fields: string[] = [];
  if (!isEmailAddress(normalized.email_address)) fields.push("email_address");
  if (!isDisplayName(normalized.display_name)) fields.push("display_name");
  if (!isFrontendLocale(normalized.preferred_language)) {
    fields.push("preferred_language");
  }
  if (!isCountryCode(normalized.resident_country)) {
    fields.push("resident_country");
  }
  return fields;
}

export interface CompleteSignupRequest {
  signup_token: HubSignupToken;
  password: NewPassword;
}

export function validateCompleteSignupRequest(
  request: CompleteSignupRequest,
): string[] {
  const fields: string[] = [];
  if (!isOpaqueToken(request.signup_token)) fields.push("signup_token");
  if (!isNewPassword(request.password)) fields.push("password");
  return fields;
}

export interface CompleteSignupResponse {
  hub_user_did: HubUserDID;
  handle: HubHandle;
}

File: typespec/hub/auth/signup.go
package auth

import (
	"github.com/vetchium/src/typespec/common"
	"github.com/vetchium/src/typespec/hub"
)

type HubSignupToken common.OpaqueToken

type RequestSignupRequest struct {
	EmailAddress      common.EmailAddress   `json:"email_address"`
	DisplayName       common.DisplayName    `json:"display_name"`
	PreferredLanguage common.FrontendLocale `json:"preferred_language"`
	ResidentCountry   common.CountryCode    `json:"resident_country"`
}

func (r *RequestSignupRequest) Normalize() {
	r.EmailAddress = common.NormalizeEmailAddress(r.EmailAddress)
	r.DisplayName = common.NormalizeDisplayName(r.DisplayName)
}

func (r RequestSignupRequest) Validate() []string {
	fields := make([]string, 0, 4)
	if !common.IsEmailAddress(r.EmailAddress) {
		fields = append(fields, "email_address")
	}
	if !common.IsDisplayName(r.DisplayName) {
		fields = append(fields, "display_name")
	}
	if !common.IsFrontendLocale(r.PreferredLanguage) {
		fields = append(fields, "preferred_language")
	}
	if !common.IsCountryCode(r.ResidentCountry) {
		fields = append(fields, "resident_country")
	}
	return fields
}

type CompleteSignupRequest struct {
	SignupToken HubSignupToken     `json:"signup_token"`
	Password    common.NewPassword `json:"password"`
}

func (r *CompleteSignupRequest) Normalize() {}

func (r CompleteSignupRequest) Validate() []string {
	fields := make([]string, 0, 2)
	if !common.IsOpaqueToken(string(r.SignupToken)) {
		fields = append(fields, "signup_token")
	}
	if !common.IsNewPassword(r.Password) {
		fields = append(fields, "password")
	}
	return fields
}

type CompleteSignupResponse struct {
	HubUserDID hub.HubUserDID `json:"hub_user_did"`
	Handle     hub.HubHandle  `json:"handle"`
}
